sklearn/tree/_tree.pyx

```diff
@@ -217,8 +214,8 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
                     impurity = splitter.node_impurity()
                     first = 0
 
-                is_leaf = (is_leaf or
-                           (impurity <= min_impurity_split))
+                # impurity can be < 0 due to rouding errors
+                is_leaf = is_leaf or impurity <= 0
```

Comment:
Should we be a bit more careful and do `impurity <= epsilon`?

Impurity euqal zero usually means a pure node, i.e. one where all samples have the same value of `y`. Are there some better checks for that or can it introduce a cheep check for pure nodes?

Reply:
I would be fine with that. If the impurity is at machine precision, it's likely the case that an additional split will not improve it and will thereafter be rejected. This would be a waste of computation that can be avoided.

Reply:
I'm totally fine with this change. Let's just agree that this still means that we imply that the best achievable impurity equals zero. This is currently the case for the implemented criteria. Being more flexible would make life easier (=more efficient) for the Poisson criterion.